File: evolution/crispr_titan.py
"""
CRISPR_Titan - Genetic Editing Engine
Implements 5 mutation strategies for genome editing.
"""
from __future__ import annotations

import logging
import math
import random
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Optional

logger = logging.getLogger(__name__)


class MutationStrategy(str, Enum):
    """Mutation strategies"""
    GAUSSIAN = "gaussian"          # TYPE 1: Structured Gaussian
    EPIGENETIC = "epigenetic"      # TYPE 2: Epigenetic
    TRANSPOSON = "transposon"      # TYPE 3: Gene jumping
    BRICOLAGE = "bricolage"        # TYPE 4: Genetic tinkering (from archive)
    SACRIFICE = "sacrifice"        # TYPE 5: Self-destructive altruism


class ChromosomeType(str, Enum):
    """Chromosome types"""
    STRUCT = "STRUCT"
    BEHAVIOR = "BEHAVIOR"
    EPI = "EPI"
    META = "META"


@dataclass(slots=True)
class Gene:
    """Individual gene representation"""
    id: int
    value: float
    chromosome: ChromosomeType
    mutable: bool
    h3k4me3: float  # Active mark
    h3k27me3: float  # Repressive mark
    dna_methylation: float

    def express(self) -> float:
        """Get gene expression level based on marks"""
        return self.value * (self.h3k4me3 - self.h3k27me3)


@dataclass(slots=True)
class Genome:
    """Editable genome representation"""
    agent_name: str
    generation: int
    fitness_history: list[float] = field(default_factory=list)
    g_struct: list[Gene] = field(default_factory=list)
    g_behavior: list[Gene] = field(default_factory=list)
    g_epi: list[Gene] = field(default_factory=list)
    g_meta: list[Gene] = field(default_factory=list)

    def clone(self) -> Genome:
        """Create a copy of this genome"""
        return Genome(
            agent_name=self.agent_name,
            generation=self.generation,
            fitness_history=self.fitness_history.copy(),
            g_struct=[Gene(
                id=g.id, value=g.value, chromosome=g.chromosome,
                mutable=g.mutable, h3k4me3=g.h3k4me3,
                h3k27me3=g.h3k27me3, dna_methylation=g.dna_methylation
            ) for g in self.g_struct],
            g_behavior=[Gene(
                id=g.id, value=g.value, chromosome=g.chromosome,
                mutable=g.mutable, h3k4me3=g.h3k4me3,
                h3k27me3=g.h3k27me3, dna_methylation=g.dna_methylation
            ) for g in self.g_behavior],
            g_epi=[Gene(
                id=g.id, value=g.value, chromosome=g.chromosome,
                mutable=g.mutable, h3k4me3=g.h3k4me3,
                h3k27me3=g.h3k27me3, dna_methylation=g.dna_methylation
            ) for g in self.g_epi],
            g_meta=[Gene(
                id=g.id, value=g.value, chromosome=g.chromosome,
                mutable=g.mutable, h3k4me3=g.h3k4me3,
                h3k27me3=g.h3k27me3, dna_methylation=g.dna_methylation
            ) for g in self.g_meta],
        )

    def get_phenotype(self) -> dict[str, float]:
        """Get phenotypic expression from genotype"""
        return {
            "num_layers": sum(1 for g in self.g_struct if g.express() > 0.5),
            "exploration_rate": sum(g.express() for g in self.g_behavior) / max(len(self.g_behavior), 1),
            "creativity_index": sum(g.h3k4me3 for g in self.g_epi) / max(len(self.g_epi), 1),
            "learning_rate": 1e-3 * sum(g.express() for g in self.g_meta) / max(len(self.g_meta), 1),
        }


@dataclass(slots=True)
class CRISPR_Titan:
    """
    CRISPR-Titan - Genetic Editing Engine
    Implements 5 mutation strategies for directed evolution.
    """
    success_rate: float = field(default=0.5)

    def __post_init__(self) -> None:
        """Initialize genetic editing engine"""
        logger.info("[CRISPR-Titan] 🧬 Moteur d'édition génique dirigée (5 opérateurs) initialisé.")

    async def edit(
        self,
        genome: Genome,
        target_gene_id: int,
        chrom: ChromosomeType,
        strategy: MutationStrategy
    ) -> Genome:
        """
        Edit a specific gene in genome according to strategy.
        Returns modified genome if fitness improves, original otherwise.
        """
        logger.debug(
            f"[CRISPR-Titan] Édition du gène {target_gene_id} sur {chrom} via {strategy}"
        )

        clone = genome.clone()
        new_value = self._get_gene_value(genome, target_gene_id, chrom)

        if strategy == MutationStrategy.GAUSSIAN:
            new_value = self._gaussian_mutation(new_value)
        elif strategy == MutationStrategy.TRANSPOSON:
            new_value = self._transposon_mutation(genome) or new_value
        elif strategy == MutationStrategy.BRICOLAGE:
            # Would resurrect from archive
            new_value = max(0, new_value - 0.1)
        elif strategy == MutationStrategy.SACRIFICE:
            new_value = 0.0
        # EPIGENETIC doesn't change base value

        if strategy != MutationStrategy.EPIGENETIC:
            self._apply_mutation(clone, target_gene_id, chrom, new_value)

        clone_fitness = await self._simulate_clone(clone)
        original_fitness = (
            genome.fitness_history[-1] if genome.fitness_history else 0.5
        )

        if clone_fitness > original_fitness or strategy == MutationStrategy.SACRIFICE:
            self._apply_mutation(genome, target_gene_id, chrom, new_value)
            self._epigenetic_mark(genome, target_gene_id, chrom, "activated")
            self.success_rate = 0.9 * self.success_rate + 0.1
            return genome
        else:
            self._epigenetic_mark(genome, target_gene_id, chrom, "repressed")
            self.success_rate = 0.9 * self.success_rate
            return genome

    def _gaussian_mutation(self, current: float, std_dev: float = 0.1) -> float:
        """Apply Gaussian mutation"""
        noise = (random.random() - 0.5) * (std_dev / 0.1)
        return max(0, min(1, current + noise))

    def _transposon_mutation(self, genome: Genome) -> Optional[float]:
        """Jump a gene from another chromosome"""
        chroms = [ChromosomeType.STRUCT, ChromosomeType.BEHAVIOR, ChromosomeType.EPI, ChromosomeType.META]
        source_chrom = random.choice(chroms)
        source_gene_id = random.randint(0, 99)
        return self._get_gene_value(genome, source_gene_id, source_chrom)

    def _apply_mutation(
        self, genome: Genome, target: int, chrom: ChromosomeType, value: float
    ) -> None:
        """Apply mutation to gene"""
        genes = self._get_chromosome(genome, chrom)
        if target < len(genes) and genes[target].mutable:
            genes[target].value = value

    def _epigenetic_mark(
        self, genome: Genome, target: int, chrom: ChromosomeType, mark_type: str
    ) -> None:
        """Add epigenetic marks"""
        genes = self._get_chromosome(genome, chrom)
        if target >= len(genes):
            return

        if mark_type == "activated":
            genes[target].h3k4me3 = 0.9
            genes[target].h3k27me3 = 0.1
        elif mark_type == "repressed":
            genes[target].h3k4me3 = 0.1
            genes[target].h3k27me3 = 0.9
        elif mark_type == "trial":
            genes[target].h3k4me3 = 0.5
            genes[target].h3k27me3 = 0.5

    def _get_chromosome(self, genome: Genome, chrom: ChromosomeType) -> list[Gene]:
        """Get chromosome by type"""
        match chrom:
            case ChromosomeType.STRUCT:
                return genome.g_struct
            case ChromosomeType.BEHAVIOR:
                return genome.g_behavior
            case ChromosomeType.EPI:
                return genome.g_epi
            case ChromosomeType.META:
                return genome.g_meta

    def _get_gene_value(
        self, genome: Genome, target: int, chrom: ChromosomeType
    ) -> float:
        """Get gene value"""
        genes = self._get_chromosome(genome, chrom)
        return genes[target].value if target < len(genes) else 0.5

    async def _simulate_clone(self, clone: Genome) -> float:
        """Simulate clone fitness (deterministic)"""
        phenotype = clone.get_phenotype()

        layer_score = 1.0 if 24 <= phenotype.get("num_layers", 0) <= 72 else 0.5
        exploration_score = 1.0 - abs(phenotype.get("exploration_rate", 0.5) - 0.6)
        creativity_score = 1.0 - abs(phenotype.get("creativity_index", 0.5) - 0.5)

        lr = phenotype.get("learning_rate", 1e-3)
        lr_score = 1.0 if 1e-4 <= lr <= 1e-2 else (0.6 if lr >= 1e-5 else 0.2)

        epi_sum = sum(min(g.express(), 1.0) for g in clone.g_struct[:50])
        epi_score = epi_sum / min(50, len(clone.g_struct))

        history_base = clone.fitness_history[-1] if clone.fitness_history else 0.5

        fitness = (
            history_base * 0.30 +
            layer_score * 0.20 +
            exploration_score * 0.15 +
            creativity_score * 0.15 +
            lr_score * 0.10 +
            epi_score * 0.10
        )

        return max(0, min(1, fitness))

    def process(self, input_data: Any) -> dict[str, Any]:
        """
        Process genetic editing request.
        CIEL compatibility method.
        """
        if isinstance(input_data, dict):
            return {
                "success_rate": self.success_rate,
                "status": "ready"
            }
        return {"success_rate": self.success_rate, "status": "ready"}

File: evolution_hydra/CRISPR_Titan.ts
import { ChromosomeType } from './TitanGenome.js';
import { ArcheDeNoe } from './ArcheDeNoe.js';
import logger from '../utils/logger.js';

export type MutationStrategy = 
  | 'gaussian'     // TYPE 1: Mutation Gaussienne Structurée
  | 'epigenetic'   // TYPE 2: Mutation Épigénétique
  | 'transposon'   // TYPE 3: Mutation Transposon (saut de gène)
  | 'bricolage'    // TYPE 4: Bricolage Génétique (arche)
  | 'sacrifice';   // TYPE 5: Mutation Auto-destructrice (altruisme)

interface EditableGene {
  id: number;
  value: number;
  chromosome: ChromosomeType | string;
  mutable: boolean;
  h3k4me3: number;
  h3k27me3: number;
  dna_methylation: number;
  express(): number;
}

interface EditableGenome {
  agentName: string;
  generation: number;
  fitnessHistory: number[];
  g_struct: EditableGene[];
  g_behavior: EditableGene[];
  g_epi: EditableGene[];
  g_meta: EditableGene[];
  clone(): EditableGenome;
  getPhenotype(): Record<string, number>;
}

export class CRISPR_Titan {
  private geneBank: ArcheDeNoe;
  private successRate: number = 0.5;

  constructor(geneBank: ArcheDeNoe) {
    this.geneBank = geneBank;
    logger.info('[CRISPR-Titan] 🧬 Moteur d\'édition génique dirigée (5 opérateurs) initialisé.');
  }

  /**
   * Édite un gène spécifique d'un génome selon une stratégie
   * Stratégies valides : gaussian, epigenetic, transposon, bricolage, sacrifice
   * Le clone est d'abord testé (simulateClone), puis appliqué si la fitness augmente.
   */
  public async edit(genome: EditableGenome, targetGeneId: number, chrom: ChromosomeType, strategy: MutationStrategy): Promise<EditableGenome> {
    logger.debug(`[CRISPR-Titan] Édition du gène ${targetGeneId} sur ${chrom} via ${strategy}`);

    // Clone pour tester l'édition
    const clone = genome.clone();
    let newValue = this.getGeneValue(genome, targetGeneId, chrom);

    switch (strategy) {
      case 'gaussian':
        newValue = this.gaussianMutation(newValue);
        break;
      case 'transposon':
        newValue = this.transposonMutation(genome) ?? newValue;
        break;
      case 'bricolage':
        const extinctGene = this.geneBank.resurrectGene(chrom, targetGeneId);
        if (extinctGene) newValue = extinctGene.value;
        break;
      case 'sacrifice':
        newValue = 0.0; // Désactivation forcée (altruisme)
        break;
      case 'epigenetic':
        // C'est juste un marquage, pas de changement de valeur de base
        break;
    }

    if (strategy !== 'epigenetic') {
      this.applyMutation(clone, targetGeneId, chrom, newValue);
    }
    
    // Test simulation (simplifié)
    const cloneFitness = await this.simulateClone(clone);
    const originalFitness = genome.fitnessHistory.length > 0 ? genome.fitnessHistory[genome.fitnessHistory.length - 1] : 0.5;

    // Décision
    if (cloneFitness > originalFitness || strategy === 'sacrifice') {
      // Succès: appliquer au vrai génome
      this.applyMutation(genome, targetGeneId, chrom, newValue);
      this.epigeneticMark(genome, targetGeneId, chrom, 'activated');
      this.successRate = 0.9 * this.successRate + 0.1;
      return genome;
    } else {
      // Échec: rollback et répression
      this.epigeneticMark(genome, targetGeneId, chrom, 'repressed');
      this.successRate = 0.9 * this.successRate;
      return genome;
    }
  }

  private gaussianMutation(current: number, stdDev = 0.1): number {
    const noise = (Math.random() - 0.5) * (stdDev / 0.1);
    return Math.max(0, Math.min(1, current + noise));
  }

  private transposonMutation(genome: EditableGenome): number | null {
    const chroms = [ChromosomeType.STRUCT, ChromosomeType.BEHAVIOR, ChromosomeType.EPI, ChromosomeType.META];
    const sourceChrom = chroms[Math.floor(Math.random() * chroms.length)];
    const sourceGeneId = Math.floor(Math.random() * 100); 
    return this.getGeneValue(genome, sourceGeneId, sourceChrom);
  }

  private applyMutation(genome: EditableGenome, target: number, chrom: ChromosomeType, value: number) {
    const genes = this.getChromosome(genome, chrom);
    if (genes[target] && genes[target].mutable) {
      genes[target].value = value;
    }
  }

  public epigeneticMark(genome: EditableGenome, target: number, chrom: ChromosomeType, markType: 'activated' | 'repressed' | 'trial') {
    const genes = this.getChromosome(genome, chrom);
    if (!genes[target]) return;

    if (markType === 'activated') {
      genes[target].h3k4me3 = 0.9;
      genes[target].h3k27me3 = 0.1;
    } else if (markType === 'repressed') {
      genes[target].h3k4me3 = 0.1;
      genes[target].h3k27me3 = 0.9;
    } else if (markType === 'trial') {
      genes[target].h3k4me3 = 0.5;
      genes[target].h3k27me3 = 0.5;
    }
  }

  private getChromosome(genome: EditableGenome, chrom: ChromosomeType): EditableGene[] {
    switch (chrom) {
      case ChromosomeType.STRUCT: return genome.g_struct;
      case ChromosomeType.BEHAVIOR: return genome.g_behavior;
      case ChromosomeType.EPI: return genome.g_epi;
      case ChromosomeType.META: return genome.g_meta;
    }
  }

  private getGeneValue(genome: EditableGenome, target: number, chrom: ChromosomeType): number {
    const genes = this.getChromosome(genome, chrom);
    return genes[target] ? genes[target].value : 0.5;
  }

  public async simulateClone(clone: EditableGenome): Promise<number> {
    // P1-A: Calcul DÉTERMINISTE basé sur le phénotype réel du clone
    const phenotype = clone.getPhenotype();

    // Score architectural : des couches dans la plage optimale (24-72) donnent plus de points
    const layerScore = phenotype['num_layers'] >= 24 && phenotype['num_layers'] <= 72 ? 1.0 : 0.5;

    // Score comportemental : exploration + créativité bien équilibrées
    const explorationScore = 1.0 - Math.abs(phenotype['exploration_rate'] - 0.6);
    const creativityScore  = 1.0 - Math.abs(phenotype['creativity_index']  - 0.5);

    // Score méta-génomique : learning_rate dans une plage saine
    const lr = phenotype['learning_rate'];
    const lrScore = lr >= 1e-4 && lr <= 1e-2 ? 1.0 : lr >= 1e-5 ? 0.6 : 0.2;

    // Score épigénétique global : moyenne des expressions actives des gènes struct
    let epiSum = 0;
    const sample = Math.min(50, clone.g_struct.length);
    for (let i = 0; i < sample; i++) epiSum += clone.g_struct[i].express();
    const epiScore = epiSum / sample;

    // Historique de fitness du clone (s'il en a un)
    const historyBase = clone.fitnessHistory.length > 0
      ? clone.fitnessHistory[clone.fitnessHistory.length - 1]
      : 0.5;

    // Pondération finale (sans aucun aléatoire)
    const fitness =
      historyBase   * 0.30 +
      layerScore    * 0.20 +
      explorationScore * 0.15 +
      creativityScore  * 0.15 +
      lrScore       * 0.10 +
      epiScore      * 0.10;

    return Math.max(0, Math.min(1, fitness));
  }
}
